bytes/bytes/raw/file_raw_repository.py

```diff
@@ -62,3 +72,63 @@ def _raw_file_path(self, raw_id: UUID, boefje_meta: BoefjeMeta) -> Path:
 
     def _index(self, raw_id: UUID) -> str:
         return str(raw_id)[: self.UUID_INDEX]
+
+
+class S3RawRepository(RawRepository):
+    def __init__(
+        self,
+        file_middleware: FileMiddleware,
+        bucket_per_org: bool,
+        s3_bucket_prefix: str = "OpenKAT-",
+        s3_bucket_name: str = "OpenKAT",
+    ) -> None:
+        self._file_middleware = file_middleware
+        self.bucket_per_org = bucket_per_org
+        self.s3_bucket_prefix = s3_bucket_prefix
+        self.s3_bucket_name = s3_bucket_name
+
+    @cached_property
+    def _s3resource(self):
+        session = BotoSession()
+        return session.resource("s3")
+
+    def get_or_create_bucket(self, organization: str):
+        # Create a bucket, and if it exists already return that instead
+        bucket_name = self.s3_bucket_name
+        if self.bucket_per_org:
+            bucket_name = f"{self.s3_bucket_prefix}{organization}"
+            try:
+                bucket = self._s3resource.create_bucket(Bucket=bucket_name)
+                bucket.wait_until_exists()
+                return bucket
+            except Exception as error:
+                logger.error("Something went wrong with creating bucket %s: %s", bucket_name, error)
+                raise error
+        return self._s3resource.Bucket(name=bucket_name)
+
+    def save_raw(self, raw_id: UUID, raw: RawData) -> None:
+        file_name = self._raw_file_name(raw_id, raw.boefje_meta)
+        contents = self._file_middleware.encode(raw.value)
+
+        logger.info("Writing raw data with id %s to s3", raw_id)
+        bucket = self.get_or_create_bucket(raw.boefje_meta.organization)
+        bucket.Object(file_name).put(Body=contents)
+
+    def get_raw(self, raw_id: UUID, boefje_meta: BoefjeMeta) -> RawData:
+        file_name = self._raw_file_name(raw_id, boefje_meta)
+        bucket = self.get_or_create_bucket(boefje_meta.organization)
+
+        try:
+            contents = bucket.Object(file_name).get()["Body"].read()
+        except self._s3resource.meta.client.exceptions.NoSuchBucket as error:
+            if error.response["Error"]["Code"] == "404":
+                raise BytesFileNotFoundException(error)
+            logger.error("Could not get file from s3: %s/%s due to %s", bucket.name, file_name, error)
+            raise error
+
+        return RawData(value=self._file_middleware.decode(contents), boefje_meta=boefje_meta)
+
+    def _raw_file_name(self, raw_id: UUID, boefje_meta: BoefjeMeta) -> str:
+        if self.bucket_per_org:
+            return str(raw_id)
+        return f"{boefje_meta.organization}/{raw_id}"
```

Comment:
Does S3 scale properly when there are many files in a "directory"? On disk we needed an extra `self._index(raw_id)` namespace but I know object storage has different properties.

Reply:
I don't have the experience of working with S3 that much. But reading off of multiple answers from stackoverflow:
["Even though your object's key might contain a '/', S3 treats the path as a plain string and puts all objects in a flat namespace."](https://stackoverflow.com/a/394505/9216830)

["The files are not really in folders, they are just strings as locations."](https://stackoverflow.com/a/30194799/9216830)

["My understanding is that S3 hashes object keys, so that lookup time for a given object is O(1)."](https://stackoverflow.com/a/48678474/9216830)

I assume that s3 storage would not benefit from using something like your `self._index(raw_id)`

Reply:
s3 is designed to handle "unlimited" amounts of files in one namespace. We dont need to worry about this.